### ISSUE TITLE
Error while updating "aws_batch_compute_environment"

### ISSUE THREAD

**Issue by shlomi-viz** [terraform-providers/terraform-provider-aws#11077]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.11
+ provider.aws v2.39.0
```
### Affected Resource(s)

* aws_batch_compute_environment
* aws_batch_job_queue

### Terraform Configuration Files

```hcl
# Compute environment
resource "aws_batch_compute_environment" "compute_environment_service" {
  compute_environment_name = "${var.compute_environment_name}"

  compute_resources {
    instance_role = "${var.instance_profile_arn}"
    spot_iam_fleet_role = "${var.spot_iam_fleet_role_arn}"

    instance_type = [
      "optimal",
    ]

    max_vcpus = 16
    desired_vcpus = 0
    min_vcpus = 0

    security_group_ids = [
      "${var.security_groups}",
    ]
    subnets = [
      "${var.subnet_ids}",
    ]

    type = "SPOT"

    bid_percentage = 100

    tags = "${var.tags}"
  }

  lifecycle {
    // Since this value will change once we submit new job we ignore it
    ignore_changes = [
      "compute_resources.0.desired_vcpus"
    ]
  }

  service_role = "${var.service_role_arn}"
  type         = "MANAGED"
}


# Job queue
resource "aws_batch_job_queue" "job_queue_service" {
  name                 = "${var.job_queue_name}"
  state                = "ENABLED"
  priority             = 1
  compute_environments = ["${aws_batch_compute_environment.compute_environment_service.arn}"]
}
```

### Debug Output

```
* module.create-service.module.create-batch.aws_batch_compute_environment.compute_environment_service (destroy): 1 error(s) occurred:

* aws_batch_compute_environment.compute_environment_service: error deleting Batch Compute Environment (batch_compute_environment_test): : Cannot delete, found existing JobQueue relationship
	status code: 400, request id: 6760af61-a579-4fc0-98d3-7b76c414a5ec
```

### Expected Behavior

1. Delete JobQueue
2. Update compute environment
3. Create new JobQueue

### Actual Behavior

1. Update compute environment - (new resource required)
2. Delete existing compute environment
3. Get error because compute environment has relationship with JobQueue

### Steps to Reproduce

1. `terraform apply`

### References

I saw some discussions on this matter but didn't found any result, example: https://github.com/terraform-providers/terraform-provider-aws/issues/2044

**Comment by theophilechevalier** [comment]
Hi,

There has been many discussions related to this case, see :
- https://github.com/terraform-providers/terraform-provider-aws/issues/2044
- https://github.com/terraform-providers/terraform-provider-aws/pull/2347
- https://github.com/terraform-providers/terraform-provider-aws/issues/3207

I had the same problem, and as explained [in this comment](https://github.com/terraform-providers/terraform-provider-aws/pull/2347#issuecomment-345292890), this is not really a bug even if I agree this is a problem!

Adding a `name_prefix` field as proposed in this [awaiting PR](https://github.com/terraform-providers/terraform-provider-aws/pull/10682/) + setting a lifecycle to `create_before_destroy = true` on `aws_batch_compute_environment` would definitely do the job. I think, however, it should be clearly explained in the documentation like it is for [`aws_launch_configuration`](https://www.terraform.io/docs/providers/aws/r/launch_configuration.html#using-with-autoscaling-groups).

**Comment by shlomi-viz** [comment]
Hey, 

Thanks for the reply. 
I saw the other discussions and most of them are closed (like the first 2 you mentioned) without a proper solution. Why do you think that this is not a bug?
should the `name_prefix` change every-time I make changes? if so, it sounds like a word-around and not a full solution. The order of deleting and rebuild resources should change. am I missing anything?

**Comment by kislyuk** [comment]
@theophilechevalier thanks for the links. This is a bug - many people unexpectedly run into this behavior with no clear idea of what went wrong. The proper solution would be for the Batch module to know that queues must be detached from compute environments before compute environments can be replaced.

The `name_prefix` workaround makes sense, however we need https://github.com/terraform-providers/terraform-provider-aws/pull/10682/ to be merged in order for that workaround to work.

**Comment by bflad** [comment]
Hi folks 👋 We just merged #10682, which adds a new `compute_environment_name_prefix` argument or completely omitting the `compute_environment_name` argument to the `aws_batch_compute_environment` resource. This will release with version 2.44.0 of the Terraform AWS Provider, Thursday next week.

This new functionality should allow a Terraform configuration supporting `aws_batch_compute_environment` replacements when combined with `create_before_destroy` (even with referencing `aws_batch_job_queue` resources), e.g.

```hcl
resource "aws_batch_compute_environment" "example" {
  compute_environment_name_prefix = "example" # or omit and the resource will generate names beginning with terraform

  # ... other configuration except compute_environment_name ...

  lifecycle {
    create_before_destroy = true
  }
}
```

If there are lingering issues with this type of setup after upgrading to version 2.44.0 of the Terraform AWS Provider when its released, please submit a new bug report and we can take a fresh look. Thanks.
